File: 0053-maximum-subarray/0053-maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        
        dp = [-10000] * (len(nums) + 1)

        # nums[i]로 시작하는 최대 subarray = max(nums[i + 1]의 최대 subarray + nums[i], nums[i])
        
        for i in range(len(nums)-1, -1, -1):
            dp[i] = max(nums[i], dp[i+1] + nums[i])

        return max(dp)

File: 0053-maximum-subarray/0053-maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        if (nums.length == 1)
            return nums[0];
        int[] dp = new int[nums.length];

        for (int i = 0; i < nums.length; i++) {
            if (i == 0) {
                dp[i] = nums[i];
                continue;
            }
            dp[i] = Math.max(dp[i-1] + nums[i], nums[i]);
        }

        return Arrays.stream(dp).max().getAsInt();
    }
}